File: pkg/strategy/pineworker/proto_mapping_test.go
package pineworker

import (
	"encoding/binary"
	"fmt"
	"math"
	"runtime"
	"testing"
	"time"

	"github.com/jftrade/jftrade-main/pkg/strategy/pineworker/pineworkerpb"
)

func TestProtoMappingRoundTripRequestAndResponse(t *testing.T) {
	request := validClientRequest()
	protoRequest := requestToProto(request)
	if protoRequest.GetJobId() != request.JobID || protoRequest.GetParams()["threshold"] != "10" {
		t.Fatalf("unexpected proto request: %#v", protoRequest)
	}
	if protoRequest.GetIncludePlots() {
		t.Fatalf("backtest request include_plots = true, want false")
	}
	batch := protoRequest.GetCandles()
	if batch.GetEncodingVersion() != candleBatchEncodingVersion || len(batch.GetPayload()) != candleBatchRecordBytes {
		t.Fatalf("unexpected candle batch: %#v", protoRequest.GetCandles())
	}
	if got := int64(binary.LittleEndian.Uint64(batch.GetPayload())); got != request.Candles[0].OpenTime {
		t.Fatalf("encoded open time = %d, want %d", got, request.Candles[0].OpenTime)
	}
	if got := math.Float64frombits(binary.LittleEndian.Uint64(batch.GetPayload()[40:])); got != request.Candles[0].Close {
		t.Fatalf("encoded close = %v, want %v", got, request.Candles[0].Close)
	}
	protoRequest.Params["threshold"] = "99"
	if request.Params["threshold"] != "10" {
		t.Fatal("request params were aliased")
	}
	request.Mode = ModeLive
	if !requestToProto(request).GetIncludePlots() {
		t.Fatal("live request include_plots = false, want true")
	}

	response := responseFromProto(&pineworkerpb.RunScriptResponse{
		JobId: "job-1",
		Plots: []*pineworkerpb.PlotOutput{{
			Name:   "plot",
			Values: []float64{1, 2},
		}},
		OrderIntents: []*pineworkerpb.OrderIntent{{
			Kind:           "entry",
			Id:             "long",
			FromEntry:      "base",
			Direction:      "long",
			Quantity:       1,
			QuantityPct:    25,
			LimitPrice:     10,
			StopPrice:      9,
			Comment:        "c",
			AlertMessage:   "a",
			DisableAlert:   true,
			BarIndex:       2,
			Time:           123,
			HasQuantity:    true,
			HasQuantityPct: true,
			HasLimitPrice:  true,
			HasStopPrice:   true,
		}},
		Alerts: []*pineworkerpb.AlertEvent{{
			Type:      "alertcondition",
			Id:        "alert-1",
			Message:   "crossed",
			Title:     "Cross",
			Frequency: "all",
			BarIndex:  2,
			Time:      123,
		}},
		VisualOutputs: []*pineworkerpb.VisualOutput{{
			Kind:        "label",
			Name:        "entry-label",
			PayloadJson: `{"text":"Long"}`,
		}},
		StrategyMetrics: &pineworkerpb.StrategyMetrics{
			BuyAndHoldPnl:             0,
			BuyAndHoldPerGain:         12.5,
			StrategyOutperformance:    -3.25,
			HasBuyAndHoldPnl:          true,
			HasBuyAndHoldPerGain:      true,
			HasStrategyOutperformance: true,
		},
		Logs:        []string{"log"},
		Warnings:    []string{"warn"},
		Diagnostics: []*pineworkerpb.Diagnostic{{Severity: "warning", Code: "x", Message: "m", Line: 1, Column: 2}},
		Metadata: &pineworkerpb.WorkerMetadata{
			WorkerId:      "worker-1",
			Version:       "0.1.0",
			PinetsVersion: "pinets",
			ScriptHash:    "script",
			DataHash:      "data",
			DurationMs:    7,
			RequestBytes:  8,
			ResponseBytes: 9,
			PeakRssBytes:  10,
		},
		Error: "worker error",
	})

	if response.JobID != "job-1" || response.Outputs[0].Values[1] != 2 || response.Plots[0].Values[0] != 1 {
		t.Fatalf("unexpected mapped response: %#v", response)
	}
	if response.Outputs[0].Kind != "plot" || response.Outputs[0].Name != response.Plots[0].Name {
		t.Fatalf("outputs were not rebuilt from plots: %#v", response.Outputs)
	}
	intent := response.OrderIntents[0]
	if intent.ID != "long" || !intent.HasStopPrice || !intent.DisableAlert {
		t.Fatalf("unexpected mapped order intent: %#v", intent)
	}
	if response.Alerts[0].ID != "alert-1" || response.Alerts[0].Frequency != "all" {
		t.Fatalf("unexpected mapped alerts: %#v", response.Alerts)
	}
	if response.VisualOutputs[0].Kind != "label" || response.VisualOutputs[0].PayloadJSON == "" {
		t.Fatalf("unexpected mapped visual outputs: %#v", response.VisualOutputs)
	}
	if response.StrategyMetrics == nil || !response.StrategyMetrics.HasBuyAndHoldPnL || response.StrategyMetrics.BuyAndHoldPerGain != 12.5 {
		t.Fatalf("unexpected mapped strategy metrics: %#v", response.StrategyMetrics)
	}
	if response.Diagnostics[0].Line != 1 || response.Metadata.Duration != 7*time.Millisecond {
		t.Fatalf("unexpected diagnostics/metadata: %#v", response)
	}
}

func TestCandleBatchEncodingGoldenVector(t *testing.T) {
	batch := candlesToProto([]Candle{{
		OpenTime: -2, CloseTime: 3, Open: 1.5, High: 2.5, Low: -0.5, Close: 2, Volume: 0,
	}})
	want := "feffffffffffffff0300000000000000000000000000f83f0000000000000440000000000000e0bf00000000000000400000000000000000"
	if got := fmt.Sprintf("%x", batch.GetPayload()); got != want {
		t.Fatalf("payload = %s, want %s", got, want)
	}
}

func BenchmarkRequestToProto200K(b *testing.B) {
	request := validClientRequest()
	request.Params = nil
	request.Candles = make([]Candle, 200_000)
	for index := range request.Candles {
		request.Candles[index] = Candle{
			OpenTime: int64(index + 1), CloseTime: int64(index + 2),
			Open: 10, High: 12, Low: 9, Close: 11, Volume: 100,
		}
	}
	b.ReportAllocs()
	b.ResetTimer()
	for range b.N {
		result := requestToProto(request)
		runtime.KeepAlive(result)
	}
}

func TestProtoMappingHandlesNilResponses(t *testing.T) {
	if response := responseFromProto(nil); response.JobID != "" {
		t.Fatalf("nil response mapped to %#v", response)
	}
	if response := responseFromProto(&pineworkerpb.RunScriptResponse{JobId: "job-1"}); response.StrategyMetrics != nil {
		t.Fatalf("response without strategy metrics mapped to %#v", response.StrategyMetrics)
	}
	if health := healthFromProto(nil); health.OK {
		t.Fatalf("nil health mapped to %#v", health)
	}
	if metadata := metadataFromProto(nil); metadata.WorkerID != "" {
		t.Fatalf("nil metadata mapped to %#v", metadata)
	}
}

func TestHealthFromProtoCopiesCapabilities(t *testing.T) {
	protoHealth := &pineworkerpb.HealthCheckResponse{
		Ok:            true,
		WorkerId:      "worker-1",
		Version:       "0.1.0",
		PinetsVersion: "pinets",
		Capabilities:  []string{"run"},
	}
	health := healthFromProto(protoHealth)
	protoHealth.Capabilities[0] = "mutated"
	if !health.OK || health.Capabilities[0] != "run" {
		t.Fatalf("unexpected health: %#v", health)
	}
}

func TestRequestToProtoDoesNotAliasNilOrMutableParams(t *testing.T) {
	request := validClientRequest()
	request.Params = nil
	if params := requestToProto(request).GetParams(); params != nil {
		t.Fatalf("nil params mapped to %#v", params)
	}

	request.Params = map[string]string{"risk": "low"}
	protoRequest := requestToProto(request)
	protoRequest.Params["risk"] = "high"
	if request.Params["risk"] != "low" {
		t.Fatalf("request params were aliased after proto mutation: %#v", request.Params)
	}
}

File: workers/pineworker/src/protoMapping.test.ts
import { describe, expect, test } from "vitest";
import { preparationOf } from "./preparedRequest";
import { candleBatchEncodingVersion, healthStatusToProto, runScriptRequestFromProto, runScriptResponseToProto } from "./protoMapping";

describe("proto mapping", () => {
  test("maps snake_case RunScriptRequest into worker request", () => {
    const request = runScriptRequestFromProto({
      job_id: "job-1",
      script_id: "script-1",
      source: `//@version=6\nstrategy("x")`,
      symbol: "US.AAPL",
      timeframe: "1",
      mode: "backtest",
      candles: {
        encoding_version: candleBatchEncodingVersion,
        payload: encodeCandles([{ openTime: 1, closeTime: 2, open: 10, high: 12, low: 9, close: 11, volume: 100 }]),
      },
      params: { threshold: 10 },
      include_plots: false,
    });

    expect(request.jobId).toBe("job-1");
    expect(request.scriptId).toBe("script-1");
    expect(request.candles[0]).toEqual({ openTime: 1, closeTime: 2, open: 10, high: 12, low: 9, close: 11, volume: 100 });
    expect(request.params).toEqual({ threshold: "10" });
    expect(request.includePlots).toBe(false);
    expect(preparationOf(request).dataHash).toHaveLength(64);
  });

  test("rejects legacy and malformed binary candle batches", () => {
    expect(() => runScriptRequestFromProto({
      job_id: "job-1",
      source: `//@version=6\nstrategy("x")`,
      symbol: "US.AAPL",
      timeframe: "1",
      mode: "backtest",
      candles: {
        open_time: [1],
      },
      params: {},
    })).toThrow("unsupported candle batch encoding version: 0");
    expect(() => runScriptRequestFromProto({
      job_id: "job-1",
      source: `//@version=6\nstrategy("x")`,
      symbol: "US.AAPL",
      timeframe: "1",
      mode: "backtest",
      candles: { encoding_version: candleBatchEncodingVersion, payload: Buffer.alloc(55) },
      params: {},
    })).toThrow("payload length 55 is not a multiple of 56");
    expect(() => runScriptRequestFromProto({
      job_id: "job-1",
      source: `//@version=6\nstrategy("x")`,
      symbol: "US.AAPL",
      timeframe: "1",
      mode: "backtest",
      candles: { encoding_version: candleBatchEncodingVersion, payload: Buffer.alloc(200_001 * 56) },
      params: {},
    })).toThrow("too many candles: 200001 > 200000");
  });

  test("decodes the shared binary golden vector", () => {
    const payload = Buffer.from("feffffffffffffff0300000000000000000000000000f83f0000000000000440000000000000e0bf00000000000000400000000000000000", "hex");
    const request = runScriptRequestFromProto({
      job_id: "golden",
      source: "strategy(\"x\")",
      symbol: "US.TEST",
      timeframe: "1",
      mode: "backtest",
      candles: { encoding_version: candleBatchEncodingVersion, payload },
    });
    expect(request.candles).toEqual([{ openTime: -2, closeTime: 3, open: 1.5, high: 2.5, low: -0.5, close: 2, volume: 0 }]);
  });

  test("maps worker response into protobuf-shaped response", () => {
    const response = runScriptResponseToProto({
      jobId: "job-1",
      outputs: [{ name: "signal", kind: "plot", values: [1] }],
      plots: [{ name: "signal", values: [1] }],
      orderIntents: [{
        kind: "entry",
        id: "long",
        direction: "long",
        quantity: 1,
        barIndex: 0,
        time: 1,
      }],
      alerts: [{
        type: "alertcondition",
        id: "alert-1",
        message: "crossed",
        title: "Cross",
        frequency: "all",
        barIndex: 0,
        time: 1,
      }],
      visualOutputs: [{
        kind: "label",
        name: "entry-label",
        payloadJson: "{\"text\":\"Long\"}",
      }],
      strategyMetrics: {
        buyAndHoldPnl: 0,
        buyAndHoldPerGain: 12.5,
        strategyOutperformance: -3.25,
        hasBuyAndHoldPnl: true,
        hasBuyAndHoldPerGain: true,
        hasStrategyOutperformance: true,
      },
      logs: ["log"],
      warnings: ["warn"],
      diagnostics: [{ severity: "info", code: "x", message: "ok" }],
      metadata: {
        workerId: "worker-1",
        version: "0.1.0",
        pineTSVersion: "mock",
        scriptHash: "script",
        dataHash: "data",
        durationMs: 1,
        requestBytes: 2,
        responseBytes: 3,
        peakRSSBytes: 4,
      },
    });

    expect(response).toMatchObject({
      job_id: "job-1",
      order_intents: [{ id: "long", has_quantity: true }],
      alerts: [{ id: "alert-1", frequency: "all", bar_index: 0 }],
      visual_outputs: [{ kind: "label", name: "entry-label", payload_json: "{\"text\":\"Long\"}" }],
      strategy_metrics: {
        buy_and_hold_pnl: 0,
        buy_and_hold_per_gain: 12.5,
        strategy_outperformance: -3.25,
        has_buy_and_hold_pnl: true,
        has_buy_and_hold_per_gain: true,
        has_strategy_outperformance: true,
      },
      metadata: { worker_id: "worker-1", pinets_version: "mock" },
      error: "",
    });
    expect(response).not.toHaveProperty("outputs");
  });

  test("maps health status", () => {
    expect(healthStatusToProto({
      ok: true,
      workerId: "worker-1",
      version: "0.1.0",
      pineTSVersion: "mock",
      capabilities: ["run"],
    })).toEqual({
      ok: true,
      worker_id: "worker-1",
      version: "0.1.0",
      pinets_version: "mock",
      capabilities: ["run"],
    });
  });
});

function encodeCandles(candles: Array<{ openTime: number; closeTime: number; open: number; high: number; low: number; close: number; volume: number }>): Buffer {
  const payload = Buffer.alloc(candles.length * 56);
  candles.forEach((candle, index) => {
    const offset = index * 56;
    payload.writeBigInt64LE(BigInt(candle.openTime), offset);
    payload.writeBigInt64LE(BigInt(candle.closeTime), offset + 8);
    payload.writeDoubleLE(candle.open, offset + 16);
    payload.writeDoubleLE(candle.high, offset + 24);
    payload.writeDoubleLE(candle.low, offset + 32);
    payload.writeDoubleLE(candle.close, offset + 40);
    payload.writeDoubleLE(candle.volume, offset + 48);
  });
  return payload;
}
